File: ios/Sources/DeckSearch.swift
import Foundation

/// Searching decks by name and by the text on their cards.
///
/// This is the obvious implementation, and it is the one that shipped.
///
/// An earlier version pre-lowercased every deck and card into one string per
/// deck, built once, on the theory that lowercasing per keystroke was the
/// expensive part. Measured on 2,000 decks and 40,000 cards it was 1.0-1.4x
/// faster per query and *slower* overall once the build cost was counted -
/// while debouncing the input, which needs no data structure at all, was up to
/// 7x. The index was deleted. See Chapter 13, and scripts/bench-search.swift.
enum DeckSearch {

    /// Decks matching the query, in their original order.
    ///
    /// An empty query returns everything, which is what a search field with
    /// nothing typed in it should show.
    static func results(for query: String, in decks: [Deck],
                        cards: (Deck) -> [Card]) -> [Deck] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !needle.isEmpty else { return decks }
        return decks.filter { deck in
            if deck.name.lowercased().contains(needle) { return true }
            return cards(deck).contains {
                $0.front.lowercased().contains(needle)
                    || $0.back.lowercased().contains(needle)
            }
        }
    }
}

File: android/app/src/main/kotlin/com/ashkanghaderi/flashcards/DeckSearch.kt
package com.ashkanghaderi.flashcards

/**
 * Searching decks by name and card text.
 *
 * The obvious implementation, for the reason Chapter 13 measured on the iOS
 * side: the index that replaced it was slower once its build cost was counted,
 * and debouncing beat both. The port keeps the conclusion, not just the code.
 */
object DeckSearch {
    fun results(query: String, decks: List<Deck>, cards: (Deck) -> List<Card>): List<Deck> {
        val needle = query.trim().lowercase()
        if (needle.isEmpty()) return decks
        return decks.filter { deck ->
            deck.name.lowercase().contains(needle) || cards(deck).any {
                it.front.lowercase().contains(needle) || it.back.lowercase().contains(needle)
            }
        }
    }
}
